### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by yashjain28**
A question came to my mind when I have suggested these tips!

Tip: If user wants to work with 
1. Another dataset(assuming it exists) in a project then it can be done by:
     var anotherDS = project.Dataset('YOUR_DATASETID'); // returns a Dataset Object !
2. Another table(assuming it exists) in a dataset then it can be done by:
     var anotherTable = project.Dataset.Table('YOUR_TABLEID');

So if a user wants to switch tables or datasets in his code, he'll have to create a complete object, so to do away with that I suggested these tips. But in turn, the code might not work for some basic functionalities like doing a project.dataset.table.list() if the user has switched datasets, the reason is obvious, the table object will not be instantiated. The discussion is on how do we get away with this?

### CODE REVIEW

File: .cbmeta [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+    "asset_refresh_dates": [],
+    "developer_email": "yaassh28@gmail.com",
+    "messaging_url": "staging.clearblade.com:1883",
+    "platform_url": "https://staging.clearblade.com",
+    "token": "-TOKOtJkphIOSDp5YERgzatMWRZd3mXu1GZ-YLJySUZAWW9rL1MtHVcBuTPfBhY48UY6WIOY17kNfid4zw=="
```
Comment (rreinold) [inline .cbmeta]: Did you notice a new system in your staging account?

File: Readme.md [hunk 2]
```diff
@@ -18,171 +18,444 @@ This is an ipm package, which contains one or more reusable assets within the ip
...
 ## API<a name="BigQuery"></a>
 
-## BigQuery(auth_key)
+## Functions
+
+<dl>
+<dt><a href="#BigQuery">BigQuery(options)</a> ⇒ <code>Object</code></dt>
+<dd><p>BigQuery object requires an API Key</p>
+</dd>
+<dt><a href="#ExampleDeleteDataset">ExampleDeleteDataset(req, resp)</a></dt>
```
Comment (rreinold) [inline Readme.md]: JSDoc all assets (Services and Libraries), but only generate markdown documentation for the core assets (Library)

Comment (rreinold) [reply Readme.md]: i.e. remove all Readme markdown documentation of the example code services

File: Readme.md [hunk 3]
```diff
@@ -18,171 +18,444 @@ This is an ipm package, which contains one or more reusable assets within the ip
...
+<dd><p>List All datasets in a project</p>
+</dd>
+<dt><a href="#ExampleUpdateDataset">ExampleUpdateDataset(req, resp)</a></dt>
+<dd><p>Updates i.e. Replaces the current dataset with the passed dataset</p>
+</dd>
+</dl>
+
+<a name="BigQuery"></a>
+
+## BigQuery(options) ⇒ <code>Object</code>
```
Comment (rreinold) [inline Readme.md]: This options object should be declared as a type (see @typedef)

File: Readme.md [hunk 4]
```diff
@@ -18,171 +18,444 @@ This is an ipm package, which contains one or more reusable assets within the ip
...
 BigQuery object requires an API Key
 
 **Kind**: global function  
+**Returns**: <code>Object</code> - Project
```
Comment (rreinold) [inline Readme.md]: Project, Table, Dataset should all also be types

File: code/services/ExampleDeleteDataset/ExampleDeleteDataset.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    };
+
+    var bQProj = BigQuery(options).initialize();
+
+    bQProj.Dataset.Delete(function (err, data) {
+        if (err) {
+            resp.error(data);
+        }
+        else {
+            resp.success(data);
```
Comment (rreinold) [inline code/services/ExampleDeleteDataset/ExampleDeleteDataset.js]: add a comment on each of these that represents the output. ex.

resp.success(data); // => {"operation":"success"}

File: code/services/ExampleGetDataset/ExampleGetDataset.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+ * Fetches the given dataset in a project 
+ * @param {Object} req
+ * @param {callback} resp - Dataset Resource object is returned!
+ */
+function ExampleGetDataset(req, resp){
+    var options = {
+        authToken: 'YOUR_BIGQUERY_AUTH_TOKEN' //(is the API key for Google BigQuery Service, https://cloud.google.com/bigquery/docs/authorization)
+        , projectID: 'gentle-impulse-161804'
+        , datasetID: 'babynames'
+        //(optional parameters)
```
Comment (rreinold) [inline code/services/ExampleGetDataset/ExampleGetDataset.js]: inline comments in object declarations are discouraged

File: code/services/ExampleInsertAllRows/ExampleInsertAllRows.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
+/**
+ * Inserts all the rows in the given table
+ * @param {Object} req     
+ *      requestBody:{
```
Comment (rreinold) [inline code/services/ExampleInsertAllRows/ExampleInsertAllRows.js]: this is invalid JSDoc, you need to use @example

File: code/services/ExampleInsertAllRows/ExampleInsertAllRows.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+                {
+                    "json":{
+                        "name":"Robert",
+                        "gender":"M",
+                        "count":10
+                    }
+                }
+            ]
+    }
+    var bQProj = BigQuery(options).initialize();
```
Comment (rreinold) [inline code/services/ExampleInsertAllRows/ExampleInsertAllRows.js]: Why don't we initialize in BigQuery's constructor?

Comment (rreinold) [reply code/services/ExampleInsertAllRows/ExampleInsertAllRows.js]: The answer to this question may help us with the main question you asked in the PR

File: package.json [hunk 9]
```diff
@@ -8,7 +9,7 @@
     "description": "BigQuery is Google's serverless, highly scalable, low cost enterprise data warehouse designed to make all your data analysts productive.",
     "domains": [],
     "gateways": [],
-    "image_url": "https://docs.clearblade.com/v/3/static/img/default_package.png",
+    "image_url": "https://imgur.com/a/coN9r",
```
Comment (rreinold) [inline package.json]: this currently points to a webpage. you need to add .png to point to the image

File: Readme.md [hunk 10]
```diff
@@ -18,171 +18,444 @@ This is an ipm package, which contains one or more reusable assets within the ip
 
 ## Setup
 
-_Add any setup instructions, such as an API Key_
+API key can be generated/accessed by going to this page: https://cloud.google.com/bigquery/docs/authorization
+ 
+Another simple way: Copy and paste the following command 'gcloud auth application-default print-access-token' in Google Cloud Platform shell!
```
Comment (rreinold) [inline Readme.md]: preface with: 'If you already have a GCloud account and the CLI set up," ...

File: Readme.md [hunk 11]
```diff
@@ -18,171 +18,444 @@ This is an ipm package, which contains one or more reusable assets within the ip
...
+**Example**  
+```js
+:- Inside of a ClearBlade service
 
-<a name="BigQuery+delete"></a>
+var project = BigQuery(options).initialize();
 
-### bigQuery.delete(projectID, datasetID, callback)
-Datasets - Delete
+Tip: If user wants to work with
```
Comment (rreinold) [inline Readme.md]: code snippets in @example should syntactically correct, this line is not

File: Readme.md [hunk 12]
```diff
@@ -18,171 +18,444 @@ This is an ipm package, which contains one or more reusable assets within the ip
...
+            * [~Get(callback)](#BigQuery..Project..Dataset..Get) ⇒
+            * [~Delete(callback)](#BigQuery..Project..Dataset..Delete)
+            * [~Insert(requestBody, callback)](#BigQuery..Project..Dataset..Insert) ⇒
+            * [~Update(requestBody, callback)](#BigQuery..Project..Dataset..Update)
+            * [~List(callback)](#BigQuery..Project..Dataset..List)
+        * [~List(callback)](#BigQuery..Project..List)
+
+<a name="BigQuery..Project"></a>
+
+### BigQuery~Project(projectID) ⇒ <code>Object</code>
```
Comment (rreinold) [inline Readme.md]: After defining the @typedef Project, this function should then return a Project object

File: Readme.md [hunk 13]
```diff
@@ -18,171 +18,444 @@ This is an ipm package, which contains one or more reusable assets within the ip
...
-| datasetID | <code>string</code> | name of the dataset you want to delete in the project |
-| callback | <code>callback</code> | callback with function signature: (err, data) |
+| requestBody | <code>Object</code> | {                         "kind": "bigquery#tableDataInsertAllRequest",                         "skipInvalidRows": boolean,                         "ignoreUnknownValues": boolean,                         "templateSuffix": string,                         "rows": [                             {                                 "insertId": string,                                 "json": {                                     (key): (value)                                 }                             }                         ]                     } |
+| callback | <code>callback</code> |  |
 
 **Example**  
 ```js
-response
-If successful, this method returns an empty response body.
+:- Inside of a ClearBlade service, after the initialization
```
Comment (rreinold) [inline Readme.md]: This is not js syntax

File: Readme.md [hunk 14]
```diff
@@ -18,171 +18,444 @@ This is an ipm package, which contains one or more reusable assets within the ip
...
+Assets:
+  Services: All services are with-respect-to a project. All the services below shows how to - 
+  1. ExampleDeleteDataset: delete a dataset.
+  2. ExampleGetDataset: get a particular dataset.    
+  3. ExampleInsertDataset: insert a dataset.
+  4. ExampleInsertAllRows: insert rows in a table of dataset.
+  5. ExampleListAllDataset: list all datasets.
+  6. ExampleUpdateDataset: updates(generally replaces) an existing dataset.
+  
 ### Code Services
```
Comment (rreinold) [inline Readme.md]: The code services above should be within this ### Code Services

File: .cbmeta (hunk 1 above)
Comment (yashjain28) [reply .cbmeta]: :(

File: Readme.md (hunk 2 above)
Comment (yashjain28) [reply Readme.md]: Okay.

File: Readme.md (hunk 4 above)
Comment (yashjain28) [reply Readme.md]: Okay, got it, custom type objects go in typedef!

File: code/services/ExampleInsertAllRows/ExampleInsertAllRows.js (hunk 7 above)
Comment (yashjain28) [reply code/services/ExampleInsertAllRows/ExampleInsertAllRows.js]: My intention was something else, but got you.

File: Readme.md (hunk 3 above)
Comment (yashjain28) [reply Readme.md]: @typedef {Object} BigQueryOptions
@property {string} authToken - YOUR_BIGQUERY_AUTH_TOKEN 
@property {string} projectID - YOUR_PROJECT_ID
@property {string} datasetID - DATASET_IN_ABOVE_PROJECT
@property {string} tableID: TABLE_IN_ABOVE_DATASET

@param {BigQueryOptions} options

Are you expecting something like this?

Comment (rreinold) [reply Readme.md]: yes

File: .cbmeta (hunk 1 above)
Comment (rreinold) [reply .cbmeta]: ;)

File: code/libraries/BigQuery/BigQuery.js [hunk 15]
```diff
@@ -1,63 +1,36 @@
...
  * 
+ * var bigQ = BigQuery(authToken, projectID);
+ *  
  * 
  */
 function BigQuery(options) {
+    var options = {
+        "authToken": authToken,
+        "projectID": projectID
+    }
```
Comment (rreinold) [inline code/libraries/BigQuery/BigQuery.js]: more concisely:
{ 
  authToken, 
  projectID 
}

File: code/libraries/BigQuery/BigQuery.js [hunk 16]
```diff
@@ -1,63 +1,36 @@
...
+
     _validateKey();
+    var cbhttp = /*Requests(); */require('https');
+
+
+    var baseUrl = "https://www.googleapis.com/bigquery/v2";
```
Comment (rreinold) [inline code/libraries/BigQuery/BigQuery.js]: constants should be CAPITAL_SNAKE_CASE

File: code/libraries/BigQuery/BigQuery.js [hunk 17]
```diff
@@ -71,329 +44,302 @@ function BigQuery(options) {
...
                         resp.error(err);
                     }
                 });
-             * 
-             */
-            function Insert(requestBody, callback) {
-                var reqOptions = _createRequestObject(datasetUrl, requestBody);
+                * 
+                */
+            function InsertAll(requestBody, callback) {
```
Comment (rreinold) [inline code/libraries/BigQuery/BigQuery.js]: javascript convention is Prototypes are PascalCase, functions are camelCase

File: code/libraries/BigQuery/BigQuery.js [hunk 18]
```diff
@@ -71,329 +44,302 @@ function BigQuery(options) {
...
-                    }
-               })
+            project.Dataset.Get(function(err, response){
+                if(!err){
+                    resp.success(response); 
+                } 
+                else{
+                    resp.error(err); 
+                }
+            });
```
Comment (rreinold) [inline code/libraries/BigQuery/BigQuery.js]: minor, but i recommend
if(err){
  resp.error(err) // this exits code service, so no need for the 'else'
}
resp.success(data)

File: code/libraries/BigQuery/BigQuery.js [hunk 19]
```diff
@@ -71,329 +44,302 @@ function BigQuery(options) {
...
+        /**
+        * Update: Updates information in an existing dataset. The update method replaces the entire dataset resource,
+        * whereas the patch method only replaces fields that are provided in the submitted dataset resource.
+        *
+        * For more Information regarding optional parameters and response structure: 
+        * https://cloud.google.com/bigquery/docs/reference/rest/v2/datasets/update
+        *
+        *
+        * @param {Object} requestBody - Should provide Dataset resource object with the format
+            {
```
Comment (rreinold) [inline code/libraries/BigQuery/BigQuery.js]: This should definitely have a @typedef with an @example

File: code/libraries/BigQuery/BigQuery.js [hunk 20]
```diff
@@ -1,63 +1,36 @@
...
  * 
  */
 function BigQuery(options) {
+    var options = {
+        "authToken": authToken,
+        "projectID": projectID
+    }
+
     _validateKey();
+    var cbhttp = /*Requests(); */require('https');
```
Comment (rreinold) [inline code/libraries/BigQuery/BigQuery.js]: calling this var `http` should be OK

File: Readme.md (hunk 2 above)
Comment (yashjain28) [reply Readme.md]: Oh ! I didn't read REMOVE ALL MARKDOWN FROM CODE SERVICES

**Review by rreinold — APPROVED** [review]

File: code/services/ExampleGetDataset/ExampleGetDataset.js (hunk 6 above)
Comment (yashjain28) [reply code/services/ExampleGetDataset/ExampleGetDataset.js]: Done.

File: package.json (hunk 9 above)
Comment (yashjain28) [reply package.json]: done

File: Readme.md (hunk 10 above)
Comment (yashjain28) [reply Readme.md]: Done

File: Readme.md (hunk 11 above)
Comment (yashjain28) [reply Readme.md]: Done

File: Readme.md (hunk 12 above)
Comment (yashjain28) [reply Readme.md]: Removed project object, issue closed

File: Readme.md (hunk 13 above)
Comment (yashjain28) [reply Readme.md]: removed, assuming the user is going to write code inside a ClearBlade service

File: code/libraries/BigQuery/BigQuery.js (hunk 19 above)
Comment (yashjain28) [reply code/libraries/BigQuery/BigQuery.js]: Done

File: code/libraries/BigQuery/BigQuery.js (hunk 18 above)
Comment (yashjain28) [reply code/libraries/BigQuery/BigQuery.js]: Done!

File: code/libraries/BigQuery/BigQuery.js (hunk 17 above)
Comment (yashjain28) [reply code/libraries/BigQuery/BigQuery.js]: Done

File: code/libraries/BigQuery/BigQuery.js (hunk 15 above)
Comment (rreinold) [reply code/libraries/BigQuery/BigQuery.js]: TODO

File: code/libraries/BigQuery/BigQuery.js (hunk 16 above)
Comment (rreinold) [reply code/libraries/BigQuery/BigQuery.js]: done

File: code/libraries/BigQuery/BigQuery.js [hunk 21]
```diff
@@ -1,301 +1,346 @@
...
+        * Updates information in an existing dataset. The update method replaces the entire dataset resource,
+        * whereas the patch method only replaces fields that are provided in the submitted dataset resource.
+        * For more Information regarding optional parameters and response structure: 
+        * https://cloud.google.com/bigquery/docs/reference/rest/v2/datasets/update
+        *
+        * @param {DatasetResource} requestBody - The DatasetResource should contain the reference dataset 
+        * @param {callback} callback Provide a function with signature: function(err, response) 
+        * 
+        * @example
+            var requestBody = {
```
Comment (rreinold) [inline code/libraries/BigQuery/BigQuery.js:182]: This example is very confusing, and not valid JavaScript. provide a valid JSON example

File: code/libraries/BigQuery/BigQuery.js [hunk 22]
```diff
@@ -1,301 +1,346 @@
...
+        *
+        * For more Information regarding optional parameters and response structure: 
+        * https://cloud.google.com/bigquery/docs/reference/rest/v2/datasets/insert             *
+        * 
+        * @param {DatasetResource} requestBody The details for the dataset to be inserted in the Project
+        * @param {callback} callback Provide a function with signature: function(err, response)
+        * 
+        * @example
+           var requestBody = {
+               "kind": "bigquery#dataset",
```
Comment (rreinold) [inline code/libraries/BigQuery/BigQuery.js]: per note above, confusing example

**Comment by yashjain28** [comment]
Made the required Changes and creating a new one if you have any more suggestions !!

**Pull request closed without merging**